runtime/lua/vim/treesitter/query.lua

```diff
@@ -8,6 +8,102 @@ Query.__index = Query
 
 local M = {}
 
+-- Some treesitter grammars extend others.
+-- We can use that to import the queries of the base language
+local base_language_map = {}
+
+-- Filter the runtime query files, the spec is like regular runtime files but in the new `queries`
+-- diretory. They resemble ftplugins, that is that you can override queries by adding things in the
+-- `queries` directory, and extend using the `after/queries` directory.
+local function filter_files(file_list)
+  local main = {}
+  local after = {}
+
+  for _, fname in ipairs(file_list) do
+    -- Only get the name of the directory containing the queries directory
+    if vim.fn.fnamemodify(fname, ":p:h:h:h:t") == "after" then
+      table.insert(after, fname)
+    -- The first one is the one with most priority
+    elseif #main == 0 then
+      main = { fname }
+    end
+  end
+
+  vim.list_extend(main, after)
+
+  return main
+end
+
+local function runtime_query_path(lang, query_name)
+  return string.format('queries/%s/%s.scm', lang, query_name)
+end
+
+local function filtered_runtime_queries(lang, query_name)
+  return filter_files(a.nvim_get_runtime_file(runtime_query_path(lang, query_name), true) or {})
+end
+
+local function get_query_files(lang, query_name)
+  local query_files = {}
+
+  local lang_files = filtered_runtime_queries(lang, query_name)
+  vim.list_extend(query_files, lang_files)
+
+  for _, base_lang in ipairs(base_language_map[lang] or {}) do
+    local base_files = filtered_runtime_queries(base_lang, query_name)
+    vim.list_extend(query_files, base_files)
+  end
+
+  return query_files
+end
+
+local function read_query_files(filenames)
+  local contents = {}
+
+  for _,filename in ipairs(filenames) do
+    vim.list_extend(contents, vim.fn.readfile(filename))
+  end
+
+  return table.concat(contents, '\n')
+end
+
+--- Returns the runtime query {query_name} for {lang}.
+--
+-- @param lang The language to use for the query
+-- @param query_name The name of the query (i.e. "highlights")
+--
+-- @return The corresponding query, parsed.
+function M.get_query(lang, query_name)
+  local query_files = get_query_files(lang, query_name)
+  local query_string = read_query_files(query_files)
+
+  if #query_string > 0 then
+    return M.parse_query(lang, query_string)
+  end
+end
+
+--- Adds an base language for a given language.
+--
+-- @param extension The extension language.
+-- @param base The language to use as a base.
+function M.add_base_language(extension, base)
+  if not base_language_map[extension] then
+    base_language_map[extension] = {}
+  end
```

Comment:
maybe add to the documentation how to create multiple base languages like we do for tsx e.g. in which order `add_base_language` should be called with the chain of base languages.